src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,414 @@
+            this.blocks = blocks;
+            this.logger = logger;
+
+            this.watcher = new Zcoin.Watching.TransactionConfirmationWatcher<Rule>
+            (
+                this,
+                blocks
+            );
+
+            this.timerLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+            this.timers = new Dictionary<uint256, Dictionary<Guid, Timer>>();
+            this.watches = new ConcurrentDictionary<Guid, TransactionWatch<Rule>>();
+        }
+
+        public async Task<Rule> AddTransactionAsync(
+            uint256 transaction,
+            int confirmation,
+            TimeSpan unconfirmedWaitingTime,
+            Callback callback,
+            TransactionConfirmationCallbackResult successData,
+            TransactionConfirmationCallbackResult timeoutData,
+            CancellationToken cancellationToken)
+        {
+            if (transaction == null)
+            {
+                throw new ArgumentNullException(nameof(transaction));
+            }
+
+            if (confirmation < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(confirmation) ,"Confirmation is less than zero.");
+            }
+
+            if (!Timer.IsValidDuration(unconfirmedWaitingTime))
+            {
+                throw new ArgumentOutOfRangeException(nameof(unconfirmedWaitingTime), "UnconfirmedWaitingTime is invalid duration.");
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            if (successData == null)
+            {
+                throw new ArgumentNullException(nameof(successData));
+            }
+
+            if (timeoutData == null)
+            {
+                throw new ArgumentNullException(nameof(timeoutData));
+            }
+
+            var rule = await this.ruleRepository.AddAsync
+            (
+                transaction,
+                confirmation,
+                unconfirmedWaitingTime,
+                successData,
+                timeoutData,
+                callback,
+                cancellationToken
+            );
+
+            await SetupTimerAsync(rule);
+
+            return rule;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var rules = await this.ruleRepository.ListAsync(cancellationToken);
+
+            foreach (var rule in rules.Where(w => !w.Completed))
+            {
+                await SetupTimerAsync(rule);
+            }
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterWriteLock();
+
+            try
+            {
+                foreach (var timerSet in timers)
+                {
+                    foreach (var timer in timerSet.Value.Where(t => t.Value != null && t.Value.Status == TimerStatus.Started))
```

Comment:
We don't need this condition anymore due to any timers will never null and already started.